src/JsonSchema/RefResolver.php

```diff
@@ -63,11 +54,36 @@ public function __construct($retriever = null)
      */
     public function fetchRef($ref, $sourceUri)
     {
-        $retriever  = $this->getUriRetriever();
-        $jsonSchema = $retriever->retrieve($ref, $sourceUri);
-        $this->resolve($jsonSchema);
+        // Get absolute uri
+        $resolver = new UriResolver();
+        $uri = $resolver->resolve($ref, $sourceUri);
+
+        // Split in to location and fragment
+        $location = $resolver->extractLocation($uri);
+        $fragment = $resolver->extractFragment($uri);
+
+        // Retrieve dereferenced schema
+        if ($location == null) {
+            $schema = end($this->schemas);
```

Comment:
I thought this was a reasonably elegant way to deal with this problem, but you are probably right - when it comes to extending it in future it does require a fairly good understanding of what is going on - which will inevitably lead to bugs.

I'm opting to use `$schemas` purely as an associative lookup, and rather than keeping track of another state variable (`$rootSchema`), I think it keeps the code a bit cleaner to simply setup the root point within the associative array.  Let me know if you think otherwise, though.